src/scripts/sync_synonym.py

```diff
@@ -0,0 +1,208 @@
+    # - filter by source
+    owned_prefix_map: PREFIX_MAP = get_owned_prefix_map(onto_config_path)
+    mappings_df['source_prefix'] = mappings_df['source_id'].apply(lambda x: x.split(':')[0])
+    mappings_df = mappings_df[mappings_df['source_prefix'].isin(owned_prefix_map.keys())]
+    del mappings_df['source_prefix']
+    # - filter obsoletes (from mapping set --> source & mondo)
+    #  - filter obsoleted in Mondo
+    mappings_df = mappings_df[~mappings_df['mondo_label'].str.startswith('obsolete')]
+    #  - filter obsoleted in source
+    source_db: SqlImplementation = get_adapter(ontology_db_path)
+    source_obsoletes: Set[CURIE] = set([x for x in source_db.obsoletes()])
+    mappings_df = mappings_df[~mappings_df['source_id'].isin(source_obsoletes)]
+    # - filter mapped but deleted IDs
+    #   If there is no label, we take that as a proxy / sign that it has been deleted. Alternative ways: entities(),
+    #   all_entity_curies(), node() (empty props), node_exists() (no: falsely says 'True', prolly cuz bad xrefs exist)
+    source_labels: List[Tuple[CURIE, str]] = [x for x in source_db.labels(mappings_df['source_id'].tolist())]
+    source_deleted_ids: Set[CURIE] = set([x[0] for x in source_labels if not x[1]])
+    mappings_df = mappings_df[~mappings_df['source_id'].isin(source_deleted_ids)]
+    # - update source labels (sometimes Mondo & source have discrepancy)
+    source_labels: Dict[CURIE, str] = {curie: label for curie, label in source_labels}
+    mappings_df['source_label'] = mappings_df['source_id'].map(source_labels)
+
+    # Query synonyms: source
+    source_df: pd.DataFrame = _query_synonyms(mappings_df['source_id'].tolist(), source_db)
+    # - add mapped mondo ids
+    source_mondo_lookup = mappings_df.set_index('source_id')['mondo_id'].to_dict()
+    source_df['mondo_id'] = source_df['curie'].map(source_mondo_lookup)
+    source_df = source_df.rename(columns={'curie': 'source_id'})[['mondo_id', 'source_id', 'synonym_scope', 'synonym']]
+    # - add labels
+    source_df = source_df.merge(mappings_df, on=['mondo_id', 'source_id'], how='left')
+
+    # Query synonyms: Mondo
+    mondo_df = pd.read_csv(mondo_synonyms_path, sep='\t').rename(columns={'?dbXref': 'source_id'})
+    mondo_df = mondo_df.rename(columns=lambda x: x.lstrip('?'))
+    # - filter by source
+    mondo_df['source_prefix'] = mondo_df['source_id'].apply(lambda x: x.split(':')[0])
+    mondo_df = mondo_df[mondo_df['source_prefix'].isin(owned_prefix_map.keys())]
+    del mondo_df['source_prefix']
+    # - filter by source previously filtered mappings_df (obsoletes & deletions)
+    mondo_df = mondo_df[mondo_df['mondo_id'].isin(set(mappings_df['mondo_id']))]  # alternative: JOIN-based solution
+    # - filter out rows which don't appear in mondo.sssom.tsv for whatever reason
+    mondo_df = mondo_df.merge(mappings_df[['mondo_id', 'source_id']], on=['mondo_id', 'source_id'], how='inner')
```

Comment:
## FYI: Filter SPARQL synonym matches not in `mondo.sssom.tsv`
- [ ] @matentzn question below

### What's going on
We have a SPARQL query on Mondo which returns: `mondo_id`, `source_id`, `synonym_scope`, `synonym`, and `synonym_type`.

However, there are some combinations of `mondo_id` & `source_id` in the results from this query which do not appear in `mondo.sssom.tsv`. 

### Question
@matentzn Any idea why this would be the case?

Here is an example of something that appears in our query results but not in `mondo.sssom.tsv`:
```obo
id: MONDO:0800250
synonym: "GRDX2" EXACT ABBREVIATION [OMIM:300351]
xref: OMIM:300351 {source="MONDO:includedEntryInOMIM"}
```

My guess is that `mondo.sssom.tsv` is filtering out any instances of `{source="MONDO:includedEntryInOMIM"}`.

### Implementation
@twhetzel Just FYI. This might be more information than you need to know.

I spotted this issue because I was getting rows with no labels when I did a left join on the SPARQL query results (`mondo_df`) to `mondo.sssom.tsv` (`mappings_df`). 

I solved this problem by filtering out all those rows by doing an inner join on `['mondo_id', 'source_id']`. Just FYI, here were two alternatives which I tried which achieve the same exact results (in OMIM, at least).

```python
# 1. Index based solution
mondo_df = mondo_df[mondo_df.set_index(['mondo_id', 'source_id']).index.isin(    mappings_df.set_index(['mondo_id', 'source_id']).index)].reset_index(drop=True)
# 2. Filter where no labels
mondo_df = mondo_df[~(mondo_df['mondo_label'].isna() | mondo_df['source_label'].isna())]
```

I think the inner join approach is easier to understand. And also I think there could be edge cases where the labels approach (2) could give unintended results. E.g. there could be situations where labels are missing, even if they appear in `mondo.sssom.tsv`.

Reply:
For (1) I looked at a handful of these and it does seem like entries where the xref is to an OMIM term and has the source value `MONDO:includedEntryInOMIM` that Mondo-OMIM mapping is not included in the "mondo.sssom.tsv" file.

For (2), can you give an example of "we have synonyms that are mapped to non-obsolete terms which don't appear in mondo.sssom.tsv"? Are these synonyms and terms in Mondo or the external source?

Reply:
The example above is such an example:
```obo
id: MONDO:0800250
synonym: "GRDX2" EXACT ABBREVIATION [OMIM:300351]
xref: OMIM:300351 {source="MONDO:includedEntryInOMIM"}
```

In this example, by "we have synonyms", I mean `synonym: "GRDX2"` (appears both in Mondo and in the source), and by "mapped to non-obsolete term", I mean `OMIM:300351`. The synonym

Reply:
> (1) My guess is that mondo.sssom.tsv is filtering out any instances of {source="MONDO:includedEntryInOMIM"}.

To answer such questions it is best to try and read the goal, no need to guess: 

https://github.com/monarch-initiative/mondo/blob/670f2f9afd27552e012c4521478771b9adc56cb9/src/ontology/mondo.Makefile#L578

It appears to me that only exact and broad matches are included in the mondo.sssom.tsv file. `xref: OMIM:300351 {source="MONDO:includedEntryInOMIM"}` does not have any qualifier (broad, exact) so it is probably just an xref, which means, neither broad nor exact.

> However, we have synonyms that are mapped to non-obsolete terms which don't appear in mondo.sssom.tsv.

This sentence should not matter and I would recommend to be really precise here to avoid confusion about the term "mapped to". I read this sentence as "There are synonyms which have provenance that point to a non-obsolete term that does not appear in mondo.sssom.tsv".

1. It does not matter what the provenance of the synonym says at all right now, as it is all over the place. This is what we are trying to fix.
2. That said, you are right that there should not be any IDs used for provenance of synonyms not in mondo.sssom.tsv. Above, we determined why this one is missing: it is neither "exact" nor "broad", so it is not part of mondo.sssom.tsv. We should only ever sync content across "exact" matches. So the real question for Trish and the curation team that arises from this is: "Should OMIM included entries by added as "exact" matches? I dont know what these "included" entries are, so I cant speak to that..

Reply:
I agree with Nico, but wanted to rephrase/clarify one thing:
> We should only ever sync content across "exact" matches.

Therefore, the Synonym Sync takes what is already in the `mondo.ssom.tsv` file as these "exact" matches to use as the starting point for the Synonym Sync work. As Nico says the existing provenance for synonyms is all over the place.

Reply:
So, we have synonyms that sometimes don't have xrefs, but we also have synonyms that don't have defined match types. I see.

Great clarification, Nico! I had forgotten how exact/broad/etc matches are represented in `.obo`, couldn't find my notes / the answer, and made the false assumption that they were probably defined somewhere and I just wasn't seeing them.

---

> We should only ever sync content across "exact" matches.

I tagged you guys on a [thread](https://github.com/monarch-initiative/mondo-ingest/pull/93#discussion_r1671386982) about that a week ago. I had just wanted to double check this was also the case for synonym sync. Question answered! Marking that thread resolved.

---

> So the real question "Should OMIM included entries by added as "exact" matches?

No, they shouldn't. I've been following/participating in [that discussion](https://github.com/monarch-initiative/omim/issues/116) as well. They're probably mostly/all narrow matches.

Reply:
@Nico We've since figured out what to do with "included", so I'm closing this.